File: 589.N-aryTreePreorderTraversal.js
/**
 * // Definition for a Node.
 * function Node(val, children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */
var preorder = function(root) {
    const result = [];
    (function dfs(node) {
        if (node) {
            result.push(node.val);
            if (node.children) {
                node.children.forEach(item => dfs(item));
            }
        }
    })(root);
    return result;
 };
/**
 * @param {Node} root
 * @return {number[]}
 */
var preorder = function(root) {
    const stack = [root];
    const result = [];
    while(stack.length) {
        const node = stack.pop();
        if (node) {
            result.push(node.val);
            if (node.children) {
                node.children.reverse().forEach(item => stack.push(item));
            }
        }
    }
    return result;
};

File: Golang/Tree/589.N-ary-tree-preorder-traversal.go
package main

type Node struct {
	Val      int
	Children []*Node
}

func postorder(root *Node) []int {
	ret := []int{}
	var helper func(node *Node)
	helper = func(node *Node) {
		if node != nil {
			ret = append(ret, node.Val)
			for _, c := range node.Children {
				helper(c)
			}
		}
	}
	helper(root)
	return ret
}
